enhancements/etcd/storage-migration-for-etcd-encryption.md

```diff
@@ -0,0 +1,122 @@
+
+We will use the upstream `kube-storage-version-migrator` (the controller-based successor to `oc adm migrate`). While upstream usage of the `kube-storage-version-migrator` is focused on migrating to new storage versions, we should be able to explicitly trigger the `kube-storage-version-migrator` when we need to migrate the encryption key used for storage, even when the storage version itself has not changed.
+
+## Goals
+
+* Create a `openshift-storage-migration-operator`, managed by the CVO, to manage the resources needed to run the `kube-storage-version-migrator`.
+* Initiate a storage migration of resources, even if the storage version has not changed.
+* Monitor the progress of storage migration of resources.
+* Enable the kube-apiserver to utilize the `kube-storage-version-migrator` when managing etcd encryption keys.
+* Inform the upstream `kube-storage-version-migrator` of our usage and our success/failure to gain influence for future decisions and create tests to ensure our use-cases.
+
+## Non-Goals
+
+* Will not provide a higher level API to the `StorageVersionMigration` api controlled by the `kube-storage-version-migrator`.
+* Will not implement any general control loops to automatically detect and then initiate a storage migration when needed.
+
+## Proposal
+
+* Introduce `openshift/cluster-openshift-storage-migration-operator` repository.
+  * Not required for bootstrapping or re-bootstrapping.
+  * No need for static pods.
+  * Will run on cluster.
+* Implement `openshift-storage-migration-operator`. 
+* Build `kube-storage-version-migrator` image.
+  * Fork `kubernetes-sigs/kube-storage-version-migrator` to `openshift`
+  * Enable CI on `openshift/kube-storage-version-migrator` repo.
+* Utilize `StorageVersionMigration` api when etcd storage encryption keys are updated.
+
+## User Stories [optional]
+
+## Implementation Details/Notes/Constraints [optional]
+
+New control loop added to api-server:
+
+```
+for {
+    if timeToMigrateToNewEncryptionKey {
+        for resource := range encryptedResources {
+            if storageVersionMigrationExists(resource) {
+                deleteStorageVersionMigration(resource)
+            }
+            createStorageVersionMigration(resource)
+        }
+        waitUntilMigrationsComplete()
+    }
+}
+```
+
+## Risks and Mitigations
+
+* kube-storage-version-migrator is currently in alpha with a clear plan to beta.
+
+* If encryption keys change while api-servers are a different levels, forcing a storage migration might render some api-servers unable to read the etcd data. To mitigate, we don’t request storage migrations unless all control planes are at the same level.
+
+* In order to insure that storage migrations initiated by other components. For etcd encryption, for example, we distribute and maintain multiple read keys allowing for a storage migration to “occur” even when not all control planes have the latest shared write key.
+
+## Design Details
+
```

Comment:
my main questions here: 
- how do the apiserver communicate the need for migration?
- how to they get progress info?
- what happens if multiple apiserver need migration?
- is migration cpu or memory bound on the migrator side or just heavy for etcd?
- do we need throttling?

Reply:
Please see updates in file.